python/ray/util/sgd/torch/training_operator.py

```diff
@@ -344,79 +547,118 @@ def validate_batch(self, batch, batch_info):
 
     def state_dict(self):
         """Override this to return a representation of the operator state.
+        Any argument passed into self.register will automatically be saved.
+        Use this method to save any additional state.
 
         Returns:
             dict: The state dict of the operator."""
         pass
 
     def load_state_dict(self, state_dict):
         """Override this to load the representation of the operator state.
+        Anything passed into self.register will automatically be loaded. Use
+        this method to load any additional state.
 
         Args:
             state_dict (dict): State dict as returned by the operator. """
         pass
 
-    @property
-    def device(self):
-        """torch.device: The appropriate torch device, at your convenience."""
-        return self._device
+    @staticmethod
+    def from_creators(model_creator, optimizer_creator, data_creator=None,
+                      scheduler_creator=None, loss_creator=None,
+                      serialize_data_creation=True):
+
+        if not (callable(model_creator) and callable(optimizer_creator)):
+            raise ValueError(
+                "Must provide a callable model_creator and optimizer_creator.")
+
+        class CreatorOperator(TrainingOperator):
+            def _validate_loaders(self, loaders):
+                assert loaders, "Loaders need to be returned in data_creator."
+                if isinstance(loaders, (tuple, list)):
+                    if len(loaders) == 1:
+                        return loaders, None
+                    elif len(loaders) == 2:
+                        return loaders
+                    else:
+                        raise ValueError(
+                            f"Number of loaders must be <= 2. Got {loaders}")
+                # No great way of checking type otherwise
+                return loaders, None
```

Comment:
Nice; the right way to do this is to define an abstract class outside and then just override class variables when invoking this function.

Reply:
^ bump, please define an abstract class outside that holds most of this implementation.

Reply:
I moved the class outside the function and am setting the creator functions via a class method which is what I think you're asking. But I'm still not sure that this is the right way to do this- shouldn't the creator functions be instance variables instead of class variables?

Reply:
Hm, not sure what you mean. You won't be able to get an Operator "instance" since they're all instantiated on the worker processes in the first place. RE: abstract class, you can do something like:


```
class AbstractOperator(TrainingOperator):
    _model_creator = None
    def _validate_and_create_model(self):
        self._model = self._model_creator()

# then in the operator, do:
@classmethod        
def from_creators(data_creator, model_creator):
   class CreatorOperator(AbstractOperator):
       _data_creator = data_creator
       _model_creator = model_creator
```

These are still class variables, but most of this code is abstracted out.

Reply:
done